bash_completionsV2.go

```diff
@@ -0,0 +1,257 @@
+
+    # Extract the directive integer at the very end of the output following a colon (:)
+    directive=${out##*:}
+    # Remove the directive
+    out=${out%%:*}
+    if [ "${directive}" = "${out}" ]; then
+        # There is not directive specified
+        directive=0
+    fi
+    __%[1]s_debug "The completion directive is: ${directive}"
+    __%[1]s_debug "The completions are: ${out[*]}"
+
+    if [ $((directive & shellCompDirectiveError)) -ne 0 ]; then
+        # Error code.  No completion.
+        __%[1]s_debug "Received error from custom completion go code"
+        return
+    else
+        if [ $((directive & shellCompDirectiveNoSpace)) -ne 0 ]; then
+            if [[ $(type -t compopt) = "builtin" ]]; then
+                __%[1]s_debug "Activating no space"
+                compopt -o nospace
+            fi
+        fi
+        if [ $((directive & shellCompDirectiveNoFileComp)) -ne 0 ]; then
+            if [[ $(type -t compopt) = "builtin" ]]; then
+                __%[1]s_debug "Activating no file completion"
+                compopt +o default
+            fi
+        fi
+    fi
+
+    if [ $((directive & shellCompDirectiveFilterFileExt)) -ne 0 ]; then
+        # File extension filtering
+        local fullFilter filter filteringCmd
+
+        # Do not use quotes around the $out variable or else newline
+        # characters will be kept.
+        for filter in ${out[*]}; do
+            fullFilter+="$filter|"
+        done
+
+        filteringCmd="_filedir $fullFilter"
+        __%[1]s_debug "File filtering command: $filteringCmd"
+        $filteringCmd
+    elif [ $((directive & shellCompDirectiveFilterDirs)) -ne 0 ]; then
+        # File completion for directories only
+
+        # Use printf to strip any trailing newline
+        local subdir=$(printf "%%s" "${out[0]}")
+        if [ -n "$subdir" ]; then
+            __%[1]s_debug "Listing directories in $subdir"
+            pushd "$subdir" >/dev/null 2>&1 && _filedir -d && popd >/dev/null 2>&1 || return
+        else
+            __%[1]s_debug "Listing directories in ."
+            _filedir -d
+        fi
+    else
+        local tab=$(printf '\t')
+        local longest=0
+        # Look for the longest completion so that we can format things nicely
+        while IFS='' read -r comp; do
+            comp=${comp%%%%$tab*}
+            if ((${#comp}>$longest)); then
+                longest=${#comp}
+            fi
+        done < <(printf "%%s\n" "${out[@]}")
+
+        local completions=()
+        while IFS='' read -r comp; do
+            if [ -z "$comp" ]; then
+                continue
+            fi
+
+            __%[1]s_debug "Original comp: $comp"
+
+            # First properly format the description string which follows a tab character if there is one
+            if [[ "$comp" == *$tab* ]]; then
+                desc=${comp#*$tab}
+                comp=${comp%%%%$tab*}
+
+                # Add the proper number of spaces to align the descriptions
```

Comment:
If the description is very long the output will jump to the next line. This makes it less readable and does not match the behavior of zsh/fish.

I recommand something like this to trim the description based on the shell width:
````
desclength=${#desc}

# $COLUMNS stores the current shell width
# remove an extra 4 because we add 2 spaces and two brackets
maxdesclength=$(( $COLUMNS - $longest - 4 ))

#make sure we have space left after the flag name
if [[ $maxdesclength -gt 0 ]]; then

	# if the description is longer than the line width result will be negative
	result=$(( $maxdesclength - $desclength ))
	if [[ $result -lt 0 ]]; then
		# trim the description here
		desc=${desc:0:$result}
	fi

	# Add the proper number of spaces to align the descriptions
	for ((i = ${#comp} ; i < $longest ; i++)); do
		comp+=" "
	done
	comp+="  ($desc)"
fi
````

Reply:
This is great! Thank you for this.
I've gone a little further and:
1. added a `...` when truncating a description
1. not pad descriptions with spaces if they would become shorter than 8 characters